service/frontend/workflow_handler.go

```diff
@@ -4984,3 +4976,80 @@ func (wh *WorkflowHandler) unaliasCreateScheduleRequestSearchAttributes(request
+	// describe to get memo, search attributes, and execution info
+	describeResponse, err := wh.historyClient.DescribeWorkflowExecution(ctx, &historyservice.DescribeWorkflowExecutionRequest{
+		NamespaceId: namespaceID.String(),
+		Request: &workflowservice.DescribeWorkflowExecutionRequest{
+			Namespace: namespaceName,
+			Execution: execution,
+		},
+	})
+	if err != nil {
+		// TODO: rewrite "workflow" in error messages to "schedule"
+		return nil, nil, err
+	}
+
+	executionInfo := describeResponse.GetWorkflowExecutionInfo()
+	var startScheduleArgs schedspb.StartScheduleArgs
+	switch executionInfo.GetStatus() {
+	case enumspb.WORKFLOW_EXECUTION_STATUS_COMPLETED:
+		reverseHistory, err := wh.GetWorkflowExecutionHistoryReverse(ctx, &workflowservice.GetWorkflowExecutionHistoryReverseRequest{
+			Namespace:       namespaceName,
+			Execution:       &commonpb.WorkflowExecution{WorkflowId: workflowID, RunId: executionInfo.Execution.RunId},
+			MaximumPageSize: 1,
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+		if len(reverseHistory.History.GetEvents()) == 0 || reverseHistory.History.GetEvents()[0].EventType != enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED {
+			return nil, nil, serviceerror.NewInternal("schedule result cannot be loaded")
+
+		}
+		err = sdk.PreferProtoDataConverter.FromPayloads(reverseHistory.History.GetEvents()[0].GetWorkflowExecutionCompletedEventAttributes().Result, &startScheduleArgs)
+		if err != nil {
+			return nil, nil, err
+		}
+
+	case enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING:
+	default:
+		return nil, nil, serviceerror.NewNotFound("schedule not found")
+	}
+	return executionInfo, &startScheduleArgs, nil
+}
+
+func (wh *WorkflowHandler) signalWithStartScheduleWorkflow(ctx context.Context, workflowID string, namespaceName string, namespaceID namespace.ID, signalWithStarRequest *workflowservice.SignalWithStartWorkflowExecutionRequest, operationName string) error {
+
+	sizeLimitError := wh.config.BlobSizeLimitError(namespaceName)
+	sizeLimitWarn := wh.config.BlobSizeLimitWarn(namespaceName)
+	for payloadSize := range []int{signalWithStarRequest.Input.Size(), signalWithStarRequest.SignalInput.Size()} {
+		if err := common.CheckEventBlobSizeLimit(
+			payloadSize,
+			sizeLimitWarn,
+			sizeLimitError,
+			namespaceID.String(),
+			workflowID,
+			"", // don't have runid yet
+			wh.metricsScope(ctx).WithTags(metrics.CommandTypeTag(enumspb.COMMAND_TYPE_UNSPECIFIED.String())),
+			wh.throttledLogger,
+			tag.BlobSizeViolationOperation(operationName),
+		); err != nil {
+			return err
+		}
+	}
+
+	// TODO: there is a rare race condition where the schedule could be running when we desribe it in patch or update, but completed when
+	// we try to send a signal to it. This leads to the "input" payload being empty, which results in creating a schedule with no spec.
+	// This race condition is rare enough that we don't need to worry about it in short term, but we should fix it eventually.
```

Comment:
I'm a little concerned about this. There's two races: 1) completed → running and 2) running → completed. Actually also 3) completed→running→completed and 4) running→completed→running.

If we do SignalWithStart:
1: sends signal, works fine
2: starts empty workflow, totally loses schedule
3: restarts workflow with out-of-date data
4: sends signal, works fine

If we do a manual signal or restart in a loop:
1: tries to restart, gets conflict, switches to signal, works fine
2: tries to send signal, fails, switches to restart, works fine
3: restarts with out of date data (this is the one that can be fixed by my new suggested primitive)
4: sends signal, works fine

Based on that summary, the manual signal or restart is more robust in more cases and never totally loses data, so I think we should do it now

Reply:
agreed and let me update this.